apollo-federation/src/query_graph/graph_path/transition.rs

```diff
@@ -26,3 +55,508 @@ impl GraphPathTriggerVariant for QueryGraphEdgeTransition {
+                    bail!("Subgraph schema unexpectedly missing metadata");
+                };
+                // TODO: The @override directive may genuinely not exist in the subgraph schema, and
+                // override_directive_definition() should arguably return a `Result<Option<_>, _>`
+                // instead, to distinguish between that case and invariant violations.
+                let Ok(override_directive_definition) = metadata
+                    .federation_spec_definition()
+                    .override_directive_definition(other_subgraph_schema)
+                else {
+                    return Ok(None);
+                };
+                let Some(application) = field_in_other_subgraph
+                    .directives
+                    .get(&override_directive_definition.name)
+                else {
+                    return Ok(None);
+                };
+                if field_subgraph.as_ref()
+                    != metadata
+                        .federation_spec_definition()
+                        .override_directive_arguments(application)?
+                        .from
+                {
+                    return Ok(None);
+                }
+                Ok(Some(other_subgraph.clone()))
+            })
+            .process_results(|iter| iter.flatten().collect())
+    }
+
+    /// For the first element of the pair, the data has the same meaning as in
+    /// `SimultaneousPathsWithLazyIndirectPaths.advance_with_operation_element()`. We also actually
+    /// need to return a `Vec` of options of simultaneous paths (because when we type explode, we
+    /// create simultaneous paths, but as a field might be resolved by multiple subgraphs, we may
+    /// have also created multiple options).
+    ///
+    /// For the second element, it is true if the result only has type-exploded results.
+    // PORT_NOTE: In the JS codebase, this was named `advancePathWithDirectTransition`.
+    #[allow(dead_code)]
+    fn advance_with_transition(
+        &self,
+        transition: &QueryGraphEdgeTransition,
+        condition_resolver: &mut impl ConditionResolver,
+        override_conditions: &EnabledOverrideConditions,
+    ) -> Result<Either<Vec<TransitionGraphPath>, UnadvanceableClosures>, FederationError> {
+        ensure!(
+            transition.collect_operation_elements(),
+            "Supergraphs shouldn't have transitions that don't collect elements",
+        );
+        let mut to_advance = Cow::Borrowed(self);
+        if let QueryGraphEdgeTransition::FieldCollection {
+            source,
+            field_definition_position,
+            ..
+        } = transition
+        {
+            let tail_weight = self.graph.node_weight(self.tail)?;
+            if let Ok(tail_type) = <QueryGraphNodeType as TryInto<
+                CompositeTypeDefinitionPosition,
+            >>::try_into(tail_weight.type_.clone())
```

Comment:
How about `try_from`?
```suggestion
            if let Ok(tail_type) = CompositeTypeDefinitionPosition::try_from(tail_weight.type_.clone())
```